### DISCUSSION TITLE
Why use the node for deploying mon obtained by canary?

### DISCUSSION THREAD

**Discussion by fengjiankui121** [rook/rook#9960] [Q&A]
The node obtained by canary is not necessarily used to deploy mon, but rook-ceph-mon-endpoints only records the node ip obtained by canary, which will cause the node ip recorded in rook-ceph-mon-endpoints to be inconsistent with the actual deployment.

#------------------The node obtained by canary is not used to deploy mon. The relevant code is as follows：

	if schedule == nil || schedule.Zone != "" {
		k8sutil.SetNodeAntiAffinityForPod(&d.Spec.Template.Spec, requiredDuringScheduling(&c.spec), v1.LabelHostname,
			map[string]string{k8sutil.AppAttr: AppName}, nil)
	} else {
		p.PodAffinity = nil
		p.PodAntiAffinity = nil
		k8sutil.SetNodeAntiAffinityForPod(&d.Spec.Template.Spec, requiredDuringScheduling(&c.spec), v1.LabelHostname,

**Comment by travisn** [comment]
@fengjiankui121 Please provide more details for what you are trying to accomplish? What settings are in your cluster.yaml? Are you trying to use host networking? The default is to not use host networking and creates services for the mon endpoints so the mon endpoints will not be tied to the host ip.

**Reply by fengjiankui121** [reply]
@travisn I use  host networking.

**Reply by travisn** [reply]
Can you provide an example of where a mon is not assigned to the correct ip with host networking? This feature has been there for a long time and we haven't had an issue reported about this.

**Reply by fengjiankui121** [reply]
@travisn There are only three nodes in the environment that can deploy mon, and this deployment expects to deploy 3 mon.The The reproduction process of duplicate mon ip in rook-ceph-endpoint is as follows:
1、Deploy rook-ceph to create mon.a, mon.b, mon.c
2、On the node of mon.a, copy /var/lib/rook/mon-a as /var/lib/rook/mon-d, and copy /var/lib/rook/mon for mon-a and mon-d -a/data/store.db is deleted, so that both mon-a and mon-d can't get up
3、Wait for the health check of rook-ceph to create the deploy of mon-d
4、Restart the operator, and wait for the operator to create a mon-d canary (this time it is in the pending state)
5、Delete /var/lib/rook/mon-d of mon-d, mon-d will be up and added to quorum
6、Restart the pod of mon-b, wait for mon-a to be deleted, and check the data information of rook-ceph-endpoint

**Reply by fengjiankui121** [reply]
![image](https://user-images.githubusercontent.com/37136232/165439587-5d565edf-f788-45e7-9685-9d9a1652e494.png)

**Reply by travisn** [reply]
I see, the operator was restarted before the reconcile of the mon failover was completed, and the mon canary chose a different node after the operator restart. The repro steps aren't going to be hit in production, but I could imagine an operator restart for some other reason to cause this. Let's convert this to an issue.

**Reply by travisn** [reply]
The discussion conversion to an issue doesn't capture the full discussion. Could you open a new issue for this? thanks

**Reply by fengjiankui121** [reply]
@travisn  issue #9961 has been created.

**Reply by fengjiankui121** [reply]
@travisn reopen the issue #9961，please

**Reply by fengjiankui121** [reply]
@travisn The restart of the operator due to environmental problems is unavoidable, so the operator should restart at any point without causing this.

**Reply by travisn** [reply]
Did [this comment](https://github.com/rook/rook/issues/9961#issuecomment-1152451998) answer the concern?
